process/slash/detector/slashResGenerator_test.go

```diff
@@ -0,0 +1,254 @@
+			Headers:       slashableHeaders,
+			SlashingLevel: slash.CalcThreatLevel(noOfHeaders),
+		}
+	}
+
+	return slashRes
+}
+
+// This function creates a noOfHeaders, which are all signed by a randomly
+// selected noOfMaliciousSigners, from allMultiSigners.
+// If noOfMaliciousSigners < minRequiredSigners in a consensus group, then
+// the rest of the signers are randomly selected from allMultiSigners
+// If allHeadersSameConsensusGroup == true, then the same consensus group
+// will sign all the headers, otherwise validators within consensus groups are shuffled
+func generateSlashableHeaders(
+	b *testing.B,
+	hasher hashing.Hasher,
+	noOfMaliciousSigners uint32,
+	noOfHeaders uint32,
+	nodesCoordinator sharding.NodesCoordinator,
+	allMultiSigners map[string]multiSignerData,
+	allHeadersSameConsensusGroup bool,
+) (headers []data.HeaderInfoHandler, signers map[string]struct{}) {
+	consensusGroupSize := len(allMultiSigners)
+	minRequiredSignatures := calcMinRequiredSignatures(consensusGroupSize)
+	bitmapSize := calcBitmapSize(consensusGroupSize)
+	maliciousSigners := selectMaliciousSigners(noOfMaliciousSigners, allMultiSigners)
+
+	round := uint64(1)
+	epoch := uint32(0)
+	headersInfo := make([]data.HeaderInfoHandler, 0, noOfHeaders)
+	randomness := []byte("rnd")
+	for i := uint32(0); i < noOfHeaders; i++ {
+		if !allHeadersSameConsensusGroup {
+			randomness = []byte(strconv.Itoa(int(i)))
+		}
+		header := createHeaderV2(uint64(i), round, epoch, randomness)
+		publicKeys, err := nodesCoordinator.GetConsensusValidatorsPublicKeys(randomness, round, core.MetachainShardId, epoch)
+		require.Nil(b, err)
+
+		headerHash, err := core.CalculateHash(marshaller, hasher, header)
+		require.Nil(b, err)
+
+		minRequiredSigners, bitmap, sigShares := selectMinRequiredSigners(b,
+			headerHash,
+			publicKeys,
+			allMultiSigners,
+			maliciousSigners,
+			minRequiredSignatures,
+			bitmapSize)
+
+		leaderPubKey := publicKeys[0]
+		leaderPrivateKey := allMultiSigners[leaderPubKey].privateKey
+		signHeader(b, header, minRequiredSigners, bitmap, sigShares, leaderPrivateKey)
+
+		headerHash, err = core.CalculateHash(marshaller, hasher, header)
+		require.Nil(b, err)
+
+		headerInfo := &coreDataMock.HeaderInfoStub{Header: header, Hash: headerHash}
+		headersInfo = append(headersInfo, headerInfo)
+	}
+
+	return headersInfo, maliciousSigners
+}
+
+func calcMinRequiredSignatures(consensusGroupSize int) uint32 {
+	// Min required signatures = 2/3 * consensusGroup + 1
+	return uint32(0.67*float32(consensusGroupSize)) + 1
+}
+
+func calcBitmapSize(consensusGroupSize int) uint32 {
+	bitmapSize := consensusGroupSize / 8
+	if consensusGroupSize%8 != 0 {
+		bitmapSize++
+	}
+	return uint32(bitmapSize)
+}
+
+func selectMaliciousSigners(noOfMaliciousSigners uint32, allMultiSigData map[string]multiSignerData) map[string]struct{} {
+	maliciousSigners := make(map[string]struct{}, noOfMaliciousSigners)
+
+	ct := uint32(0)
+	for pubKey := range allMultiSigData {
```

Comment:
1. Regarding first part, unfortunately it is not possible since `allMultiSigData` is a `[string]multiSignerData`, therefore in this case:
- `ct` would be `key->str` and
- `pubKey` would be `multiSignerData`
2.
>also take care that noOfMaliciousSigners is below len(allMultiSigData)

There is no need for this check. If `noOfMaliciousSigners` > ` len(allMultiSigData)`, code will not panic, it would simply select  `noOfMaliciousSigners` = ` len(allMultiSigData)`

Reply:
1. Right, for some reason I was thinking it was an array.